kernel.c

```diff
@@ -0,0 +1,278 @@
+    init_interrupt_vector();
+
+    // Initialize page tables for Region 0 and 1 (Checkpoint 8.2.2)
+    // This function should also set up the initial kernel break address.
+    init_page_table((int)_first_kernel_text_page, (int)_first_kernel_data_page, (int)_orig_kernel_brk_page);
+
+    // Enable virtual memory - a critical transition point (Checkpoint 8.2.2)
+    // After this call, all memory addresses are interpreted as virtual rather than physical.
+    enable_virtual_memory();
+
+    // Create the idle process - runs when no other process is ready (Checkpoint 8.2.4)
+    // The create_idle_process function should handle setting up idle's Region 1 page table,
+    // kernel stack frames, and configuring its UserContext (PC to DoIdle, SP to user stack top).
+    pcb_t *idle_process = create_idle_process(uctxt);
+    if (!idle_process) {
+        TracePrintf(0, "ERROR: Failed to create idle process\n");
+        Halt(); // System needs at least one process to function; halt if idle creation fails.
+    }
+
+    // Set the global current_process to the newly created idle process.
+    // This makes idle the first process that the system considers 'running' or ready to run.
+    current_process = idle_process;
+
+    // Return to the idle process's user context (Checkpoint 8.2.4)
+    // This modifies the uctxt pointer provided by the hardware,
+    // so when KernelStart returns, the machine enters user mode and executes DoIdle.
+    TracePrintf(0, "Leaving KernelStart, returning to idle process (PID %d)\n", idle_process->pid);
+    *uctxt = idle_process->user_context;
+
+    return; // Return to user mode, entering the idle loop.
+}
+
+
+void init_interrupt_vector(void) {
+    // Allocate memory for interrupt vector table - array of function pointers
+    // The hardware uses this table to determine which function to call for each trap type
+    int *vector_table = (int *)malloc(sizeof(int) * TRAP_VECTOR_SIZE);
+    if (!vector_table) {
+        TracePrintf(0, "ERROR: Failed to allocate interrupt vector table\n");
+        Halt();  // Critical error - interrupt handling is essential
+    }
+
+    // Set up handlers for each trap type by populating the vector table
+    // Each entry contains the address of the function to handle that specific trap
+    vector_table[TRAP_KERNEL] = (int)kernel_handler;        // System calls from user processes
+    vector_table[TRAP_CLOCK] = (int)clock_handler;          // Timer interrupts for scheduling
+    vector_table[TRAP_ILLEGAL] = (int)illegal_handler;      // Illegal instruction exceptions
+    vector_table[TRAP_MEMORY] = (int)memory_handler;        // Memory access violations and stack growth
+    vector_table[TRAP_MATH] = (int)math_handler;            // Math errors like division by zero
+    vector_table[TRAP_TTY_RECEIVE] = (int)receive_handler;   // Terminal input available
+    vector_table[TRAP_TTY_TRANSMIT] = (int)transmit_handler; // Terminal output complete

```

Comment:
for the remaining vector table entries, you should set them to a function you create for undefined traps (which can just print and halt the machine)